### ISSUE TITLE
Tidyall/Git fails with copied files

### ISSUE THREAD

**Issue by nigelm** [houseabsolute/perl-code-tidyall#81]
When a file within a git repo has been copied (and possibly edited), and is shown appropriately within `git status`:-

    $ git status
    On branch master
    Your branch is up-to-date with 'origin/master'.
    Changes to be committed:
    (use "git reset HEAD <file>..." to unstage)

        copied:     lib/Redcentric/CDR/Split/Split.pm -> lib/Redcentric/CDR/Split/Role/Split.pm
        modified:   lib/Redcentric/CDR/Split/Split.pm

Then `tidyall -g` fails to parse the status info correctly:-

    $ tidyall -g
    Use of uninitialized value $mode in pattern match (m//) at /Users/nigel/.perlgitrel/lib/perl5/Code/TidyAll/Git/Util.pm line 60.
    Use of uninitialized value $mode in pattern match (m//) at /Users/nigel/.perlgitrel/lib/perl5/Code/TidyAll/Git/Util.pm line 65.
    Use of uninitialized value $mode in pattern match (m//) at /Users/nigel/.perlgitrel/lib/perl5/Code/TidyAll/Git/Util.pm line 60.
    Use of uninitialized value $mode in pattern match (m//) at /Users/nigel/.perlgitrel/lib/perl5/Code/TidyAll/Git/Util.pm line 65.
    Use of uninitialized value $mode in pattern match (m//) at /Users/nigel/.perlgitrel/lib/perl5/Code/TidyAll/Git/Util.pm line 60.
    Use of uninitialized value $mode in pattern match (m//) at /Users/nigel/.perlgitrel/lib/perl5/Code/TidyAll/Git/Util.pm line 65.
    [checked] lib/Redcentric/CDR/Split/Role/Split.pm

Have updated to `Code-TidyAll-0.63`.  Started to look at this myself, but am not too familiar with the porcelain output or what you are trying to do there...

Git version is `git version 2.13.0`

The output of the git status command used within the Git/Util.pm file is:-

    $ git status --porcelain -z -uno|od -c
    0000000    C           l   i   b   /   R   e   d   c   e   n   t   r   i
    0000020    c   /   C   D   R   /   S   p   l   i   t   /   R   o   l   e
    0000040    /   S   p   l   i   t   .   p   m  \0   l   i   b   /   R   e
    0000060    d   c   e   n   t   r   i   c   /   C   D   R   /   S   p   l
    0000100    i   t   /   S   p   l   i   t   .   p   m  \0   M           l
    0000120    i   b   /   R   e   d   c   e   n   t   r   i   c   /   C   D
    0000140    R   /   S   p   l   i   t   /   S   p   l   i   t   .   p   m
    0000160   \0
    0000161

Its obviously a pretty rare failure mode - I have not previously encountered this one.

Thanks for working on this - really useful package!

**Comment by autarch** [comment]
I've never gotten a "copied" status. How do you do that? Do you just copy a file that's being tracked in the repo to a new location and then `git add` it?

**Comment by nigelm** [comment]
Thats basically how I did it...

The specifics were I was using Visual Studio Code editor and used copy/paste in the file browser to replicate the file...  VS Code is git aware so there is a chance it did more under the covers, but I don't think thats it...

The info in https://stackoverflow.com/questions/2673431/git-commit-message-showing-file-as-copied-instead-of-modified-what-are-the-impl appears to say that this is deep git magic happening...

In my case (as is probably obvious by the naming) I was pushing some functionality into a common role before reuse, so the `Role/Split.pm` was very similar to the original file (which had been gutted into a fairly small shim).

**Comment by autarch** [comment]
I cannot make a file show up as copied from my manual testing. This makes fixing the bug fairly challenging ;)

If you can give me a recipe for doing so I'm sure I could fix it.

**Comment by adherzog** [comment]
Here's what I did (using this repo) to generate a copied status:

```
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean

$ cp LICENSE LICENSE2
$ tail LICENSE > LICENSE
$ git add LICENSE LICENSE2
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   LICENSE
        copied:     LICENSE -> LICENSE2
```

I think the 2 important parts are 1) the source file has to be modified and staged, and 2) the source file has to be _heavily_ modified - removing just a handful of lines wasn't enough for git to report it as `copied`.

**Comment by autarch** [comment]
Thanks, @adherzog! That let me write a test. Turns out this was an incredibly trivial fix.

**Comment by autarch** [comment]
Fixed in 0.65
